File: nikola.cpp
#include<bits/stdc++.h>

using namespace std;

typedef pair<int, int> II;
typedef pair<int, pair<int,int>> III;
typedef vector<int> VI;
typedef vector<II> VII;
typedef vector<VI> VVI;
typedef vector<VII> VVII;
#define INF 1e9

struct hash_pair {
    size_t operator()(const II& p) const {
        return hash<int>{}(p.first) ^ hash<int>{}(p.second);
    }
};

int main()
{
    int N, a;
    int m = INF;
    scanf("%d", &N);
    VVII g(N, VII());
    unordered_map<II, int, hash_pair> dist;

    VI cost(N, INF);
    for (int i=0; i<N; i++){
        scanf("%d", &a);
        cost[i] = a;
    }

    priority_queue<III, vector<III>, greater<III>> pq;
    dist[II(0,0)] = 0;
    pq.push(III(0, II(0, 0)));
    while (!pq.empty()) {
        III front = pq.top(); pq.pop();
        int d = front.first;
        int u = front.second.first, lastJump = front.second.second;
        if (d > dist[II(u,lastJump)]) continue;
        VI next = {abs(lastJump)+1, -abs(lastJump)};
        for (int nextJump : next) {
            if (u+nextJump >= N || u+nextJump < 0) continue;
            int &v = dist[II(u+nextJump,nextJump)];
            if (v == 0) { v = INF; };
            if (dist[II(u,lastJump)] + cost[u+nextJump] < dist[II(u+nextJump,nextJump)]) {
                dist[II(u+nextJump,nextJump)] = dist[II(u,lastJump)] + cost[u+nextJump];
                pq.push(III(dist[II(u+nextJump,nextJump)], II(u+nextJump, nextJump)));
                if (nextJump > 0 && u+nextJump == N-1) {
                    m = min(m, dist[II(u+nextJump,nextJump)]);
                    printf("%d\n", m);
                    return 0;
                }
            }
        }
    }
    printf("%d\n", m);
}

File: nikola.py
from sys import stdin
from queue import PriorityQueue

INF = int(1e2)
N = int(stdin.readline())
C = [int(n) for n in stdin.readlines()]

pq = PriorityQueue()
pq.put((0, 0, 0))

dist = {(0, 0): 0}
print(C)
print(dist)

while not pq.empty():
    # print([min(x, 100) for x in dist])
    cost, current, jump = pq.get()
    print(cost, current, jump)
    if dist[(current,jump)] < cost:
        continue
    if current == N-1:
        break
    nxt = current - jump
    if 0 <= nxt and nxt < N:
        print('back')
        if (nxt, jump) not in dist or dist[(nxt,jump)] + C[nxt] < dist[(nxt,jump)]:
            dist[(nxt,jump)] = dist[(nxt,jump)] + C[nxt]
            pq.put((dist[(nxt,jump)], nxt, jump))
    nxt = current + jump + 1
    if 0 <= nxt and nxt < N:
        print('nxt')
        if (nxt, jump) not in dist or dist[(nxt,jump+1)] + C[nxt] < dist[(nxt,jump+1)]:
            print('adsf')
            dist[(nxt,jump+1)] = cost + C[nxt]
            pq.put((dist[(nxt,jump)], nxt, jump+1))

print(dist)
